lib/nerves_hub_web/channels/extensions_channel.ex

```diff
@@ -0,0 +1,159 @@
+defmodule NervesHubWeb.ExtensionsChannel do
+  use Phoenix.Channel
+
+  alias Phoenix.Socket.Broadcast
+
+  require Logger
+
+  @impl Phoenix.Channel
+  def join("extensions", extension_versions, socket) do
+    extensions = parse_extensions(socket.assigns.device, extension_versions)
+    socket = assign(socket, :extensions, extensions)
+
+    attach_list = for {key, %{attach?: true}} <- extensions, do: key
+
+    if length(attach_list) > 0 do
+      send(self(), :init_extensions)
+    end
+
+    topic = "device:#{socket.assigns.device.id}:extensions"
+    NervesHubWeb.DeviceEndpoint.subscribe(topic)
+
+    {:ok, attach_list, socket}
+  end
+
+  defp parse_extensions(
+         %{extensions: device_extensions, product: %{extensions: product_extensions}},
+         extension_versions
+       ) do
+    allowed_extensions =
+      product_extensions
+      |> Map.from_struct()
+      |> Enum.filter(fn {extension, enabled?} ->
+        enabled? == true and Map.get(device_extensions, extension) != false
+      end)
+      |> Enum.map(&elem(&1, 0))
```

Comment:
```suggestion
      for {extension, true} <- Map.from_struct(product_extensions),
          {^extension, device_enabled?} <- Map.from_struct(device_extensions,
          device_enabled? != false,
          do: extension
```
Slight optimization. Less looping, more direct filtering. Gets you list of extensions only allowed at product _and_ device

Reply:
I find for comprehensions very messy to parse visually. I typically prefer a pipeline even if it is less efficient. I'll take it though :)